ethernet/w5500_spi/w5500_spi.c

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright (c) 2020 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include "pico/stdlib.h"
+#include "pico/binary_info.h"
+#include "hardware/spi.h"
+#include "w5500.h"
+#include "wizchip_conf.h"
+#include "loopback.h"
+
+
+/* Example code to talk to send and recieve data from the W5500 chip.
+
+   NOTE: Ensure the device is capable of being driven at 3.3V NOT 5V. The Pico
+   GPIO (and therefor SPI) cannot be used at 5v.
+
+   Connections on Raspberry Pi Pico board and a generic wiznet board, other
+   boards may vary.
+   GPIO 2 (physical pin 4) SCK/spi0_sclk -> CLK on wiznet board
+   GPIO 3 (physical pin 5) MOSI/spi0_tx -> MOSIon wiznet board
+   GPIO 4 (physical pin 6) MISO/spi0_rx-> MISO on wiznet board
+   GPIO 5 (physical pin 7) Chip select -> NCS on wiznet board
+   GPIO 6 (physical pin 9) Reset/GPIO -> Reset on wiznet board
+   3.3v (physical pin 36) -> VCC on wiznet board
+   GND (physical pin 38)  -> GND on wiznet board
+
+   Note: SPI devices can have a number of different naming schemes for pins. See
+   the Wikipedia page at https://en.wikipedia.org/wiki/Serial_Peripheral_Interface
+   for variations.
+
+*/
+
+#define PIN_SCK  2
+#define PIN_MOSI 3
+#define PIN_MISO 4
+#define PIN_CS   5
+#define PIN_RST  6
+
+#define SPI_PORT spi0
+#define READ_BIT 0x00
+#define TCP_SOCKET 0
+
+wiz_NetInfo gWIZNETINFO = { .mac = {0x00, 0x08, 0xdc, 0x78, 0x91, 0x71},
+                            .ip = {192, 168, 0, 15},
+                            .sn = {255, 255, 255, 255},
+                            .gw = {192, 168, 0, 1},
+                            .dns = {168, 126, 63, 1},
+                            .dhcp = NETINFO_STATIC};
+#define ETH_MAX_BUF 2048
+
+unsigned char ethBuf0[ETH_MAX_BUF];
+void print_network_information(void);
+
+
+static inline void cs_select() {
+    asm volatile("nop \n nop \n nop");
+    gpio_put(PIN_CS, 0);  // Active low
+    asm volatile("nop \n nop \n nop");
+}
+
+static inline void cs_deselect() {
+    asm volatile("nop \n nop \n nop");
+    gpio_put(PIN_CS, 1);
+    asm volatile("nop \n nop \n nop");
+}
+
+static void wiznet_reset() {
+    gpio_set_dir(PIN_RST, GPIO_OUT);
+    gpio_put(PIN_RST, 0);
+    sleep_ms(100);
+    gpio_put(PIN_RST, 1);
+    sleep_ms(100);
+    bi_decl(bi_1pin_with_name(PIN_RST, "W5500 RST"));
```

Comment:
Binary stuff should go in main, imo. Functionally I believe it's the same when using picotool, but there's no need to call it every time we reset the chip and all the other `bi_decl`s are in main anyway.